relay-sampling/src/lib.rs

```diff
@@ -809,79 +839,184 @@ impl Default for SamplingMode {
+        //
+        // An example of the aforementioned case:
+        // /hello -> /world -> /transaction belong to trace_id = abc
+        // * /hello has uniform rule with 0.2 sample rate which will match all the transactions of the trace
+        // * each project has a single transaction rule with different factors (2, 3, 4)
+        //
+        // 1. /hello is matched with a transaction rule with a factor of 2 and uses as seed abc -> 0.2 * 2 = 0.4 sample rate
+        // 2. /world is matched with a transaction rule with a factor of 3 and uses as seed abc -> 0.2 * 3 = 0.6 sample rate
+        // 3. /transaction is matched with a transaction rule with a factor of 4 and uses as seed abc -> 0.2 * 4 = 0.8 sample rate
+        //
+        // We can see that we have 3 different samples rates but given the same seed, the random number generated will be the same.
+        let event_id = event.id.value();
+
+        if has_matched_trace_rule {
+            if let Some(dsc) = dsc {
+                return Some(dsc.trace_id);
             }
+        }
 
-            if !rule.condition.matches(sampling_context, ip_addr) {
-                return None;
-            }
+        if let Some(event_id) = event_id {
+            return Some(event_id.0);
+        }
 
-            rule.is_active(now)
-        })
+        None
+    }
+
+    pub fn has_unsupported_rules(&self) -> bool {
+        !self.rules_v2.iter().all(SamplingRule::supported)
     }
 
-    /// Get the first rule of type [`RuleType::Transaction`] or [`RuleType::Error`] whose conditions
-    /// match the given event.
+    /// Matches an event and/or dynamic sampling context against the rules of the sampling configuration.
+    ///
+    /// The multi-matching algorithm used iterates by collecting and multiplying factor rules until
+    /// it finds a sample rate rule. Once a sample rate rule is found, the final sample rate is
+    /// computed by multiplying it with the previously accumulated factors.
     ///
-    /// The rule type to filter by is inferred from the event's type.
-    pub fn get_matching_event_rule(
+    /// The default accumulated factors equal to 1 because it is the identity of the multiplication
+    /// operation, thus in case no factor rules are matched, the final result will just be the
+    /// sample rate of the matching rule.
+    ///
+    /// In case no sample rate rule is matched, we are going to return a None, signaling that no
+    /// match has been found.
+    pub fn match_against_rules(
         &self,
         event: &Event,
+        dsc: Option<&DynamicSamplingContext>,
         ip_addr: Option<IpAddr>,
         now: DateTime<Utc>,
-    ) -> Option<ActiveRule> {
-        let ty = if let Some(EventType::Transaction) = &event.ty.0 {
-            RuleType::Transaction
-        } else {
-            RuleType::Error
-        };
+    ) -> Option<SamplingConfigMatchResult> {
+        let mut matched_rule_ids = vec![];
+        let mut has_matched_trace_rule = false;
```

Comment:
Instead of storing a boolean and using `get_seed`, I would just

```rust
let mut seed = event.id.value();
```

and then overwrite it with the DSC seed further down if necessary.

Reply:
I made my implementation under the assumption of having trace and transaction rules mixed, as shown in the tests. For this reason I use a boolean to keep track of any seen trace.

Reply:
What I mean is that on line 1000, instead of flipping a boolean, you could overwrite the seed directly.

Reply:
Oh yeah sorry, I get now what you mean. Will do that, sorry for the confusion.